Fix unindent in doc comments

GuillaumeGomez opened the pull request:
Fixes #70732

r? @jyn514

Nemo157 commented:
How come this doesn't break <https://github.com/QuietMisdreavus/rust/blob/master/src/test/rustdoc/issue-42760.rs>, which was explicitly testing for the old behaviour?

jyn514 commented on src/librustdoc/passes/unindent_comments.rs:
```diff
@@ -35,65 +34,72 @@ impl clean::Attributes {
+        return;
+    }
+    min_indents.sort_unstable();
+    let min_indent = min_indents[0];
```
Why did all this have to change? From what I see you don't use `min_indents` after this.

GuillaumeGomez replied:
I needed to get the `min_indent` based on all the doc fragments, not just on one.

jyn514 commented on src/librustdoc/passes/unindent_comments.rs:
```diff
@@ -35,65 +34,72 @@ impl clean::Attributes {
+    let min_indent = min_indents[0];
+
+    for (index, fragment) in docs.iter_mut().enumerate() {
+        let lines = fragment.doc.lines().collect::<Vec<&str>>();
```
```suggestion
        let lines: Vec<_> = fragment.doc.lines().collect();
```

jyn514 commented on src/test/rustdoc-ui/intra-links-warning.stderr:
```diff
@@ -96,8 +96,8 @@ LL | | /// [error]
            
-           [error]
-            ^^^^^
+            [error]
+             ^^^^^
```
Is this change intentional? It looks like this is no longer stripping leading whitespace, which doesn't seem right.

GuillaumeGomez replied:
Hum... It should definitely take into account the kind of the doc fragment. Good point!

jyn514 commented on src/librustdoc/passes/unindent_comments.rs:
```diff
@@ -35,65 +34,72 @@ impl clean::Attributes {
+        let lines = fragment.doc.lines().collect::<Vec<&str>>();
+
+        if !lines.is_empty() {
+            let mut unindented =
+                if index == 0 { vec![lines[0].trim_start().to_string()] } else { vec![] };
+            unindented.extend_from_slice(
+                &lines[if index == 0 { 1 } else { 0 }..]
```
I don't understand what this is doing, can you add a comment?

GuillaumeGomez replied:
Actually, I think that most of this code can be simplified now.

GuillaumeGomez commented:
This test only check that there is a title, not that the indent isn't broken as far as I can tell?

GuillaumeGomez commented:
@Nemo157 Well, for me it seems like it should work because sugared doc comments should be handled a bit differently than the others (they need to have one extra whitespace after all).

@jyn514 Thanks to your remarks, I greatly improved the code. :)

jyn514 commented on src/librustdoc/passes/unindent_comments.rs:
```diff
@@ -35,52 +34,64 @@ impl clean::Attributes {
-        }
-    });
+            }) + if fragment.kind == DocFragmentKind::SugaredDoc { 0 } else { 1 }
+        })
+        .collect::<Vec<_>>();
+    min_indents.sort_unstable();
+    let min_indent = min_indents[0];
```
I was suggesting something like this:
```suggestion
        })
        .min();
```

GuillaumeGomez replied:
:o That's awesome!

jyn514 commented on src/librustdoc/passes/unindent_comments.rs:
```diff
@@ -35,52 +34,64 @@ impl clean::Attributes {
+        let lines = fragment.doc.lines().collect::<Vec<&str>>();
+
+        if !lines.is_empty() {
+            let min_indent = if fragment.kind == DocFragmentKind::SugaredDoc {
+                min_indent
+            } else if min_indent > 0 {
+                min_indent - 1
+            } else {
+                min_indent
+            };
```
```suggestion
            let min_indent = if fragment.kind != DocFragmentKind::SugaredDoc && min_indent > 0 {
                min_indent - 1
            } else {
                min_indent
            };
```

Nemo157 commented:
> @Nemo157 Well, for me it seems like it should work because sugared doc comments should be handled a bit differently than the others (they need to have one extra whitespace after all).

But they don't, you can have 
```rust
///# Documentation
///
///Without leading whitespace
```
just fine, it's just convention to have a space between the doc-comment marker and the docs.

GuillaumeGomez commented:
Well, glad I treated that correctly then. Thanks for the extra information! :)

GuillaumeGomez commented:
Hum, actually I might need to make a small update in the code. :3

pickfire commented on src/librustdoc/passes/unindent_comments.rs:
```diff
@@ -35,65 +34,69 @@ impl clean::Attributes {
-                    false
+                    saw_first_line = true;
+                    let mut whitespace = 0;
+                    line.chars().all(|char| {
+                        // Compare against either space or tab, ignoring whether they
+                        // are mixed or not
+                        if char == ' ' || char == '\t' {
+                            whitespace += 1;
+                            true
+                        } else {
+                            false
+                        }
+                    });
```
We could probably simplify this a bit.
```suggestion
                    // Compare against either space or tab, ignoring whether they are
                    // mixed or not
                    let whitespace = line.chars().take_while(|c| c == ' ' || c == '\t').count();
```

GuillaumeGomez replied:
:+1:

jyn514 commented on src/librustdoc/passes/unindent_comments/tests.rs:
```diff
@@ -1,26 +1,40 @@
+    }]
+}
+
+// We use a macro to keep the line in case of a test failure.
+macro_rules! run_test {
```
Can you make this a function with `track_caller` instead?

jyn514 commented on src/test/rustdoc/unindent.rs:
```diff
@@ -0,0 +1,27 @@
+#![feature(no_core)]
+
+#![crate_name = "foo"]
+#![no_std]
+#![no_core]
```
Any reason to make this no_std and no_core?

GuillaumeGomez replied:
Woups, it was just to have less logs when running my tests. I removed them.

pickfire commented on src/librustdoc/passes/unindent_comments.rs:
```diff
@@ -35,65 +34,85 @@ impl clean::Attributes {
+            };
+
+            let mut iter = lines.iter();
+            let mut result = if !first_ignored {
```
What is `result`?

pickfire commented on src/librustdoc/passes/unindent_comments.rs:
```diff
@@ -35,65 +34,85 @@ impl clean::Attributes {
+                Vec::new()
+            };
+            result.extend_from_slice(
+                &iter
+                    .map(|&line| {
+                        if line.chars().all(|c| c.is_whitespace()) {
+                            line.to_string()
+                        } else {
+                            assert!(line.len() >= min_indent);
+                            line[min_indent..].to_string()
+                        }
+                    })
+                    .collect::<Vec<_>>(),
```
I believe this could be separated into it's own line already, at least to make it easier to understand what was being extended.

pickfire commented on src/librustdoc/passes/unindent_comments.rs:
```diff
@@ -35,65 +34,85 @@ impl clean::Attributes {
-                if char == ' ' || char == '\t' {
-                    whitespace += 1;
-                    true
+    let add = if !docs.windows(2).all(|arr| arr[0].kind == arr[1].kind)
+        && docs.iter().any(|d| d.kind == DocFragmentKind::SugaredDoc)
+    {
+        // In case we have a mix of sugared doc comments and "raw" ones, we want the sugared one to
+        // "decide" how much the minimum indent will be.
+        1
+    } else {
+        0
+    };
```
```suggestion
    let sugared_doc = if !docs.windows(2).all(|arr| arr[0].kind == arr[1].kind)
        && docs.iter().any(|d| d.kind == DocFragmentKind::SugaredDoc);
    // In case we have a mix of sugared doc comments and "raw" ones, we want the sugared one to
    // "decide" how much the minimum indent will be.
    let add = if sugared_doc { 1 } else { 0 };
```

jyn514 commented:
We're discussing this [on discord](https://discord.com/channels/442252698964721669/459149231702278154/771053580656312320), but either way the following test cases we came up with should be added:

```rust
///no whitespace
#[doc = " lol"]
```
```
///    4 whitespaces!
#[doc = "something"]
```

Additionally, note that the fixed version of https://github.com/rust-lang/rust/issues/70732 in the test case uses *4* spaces instead of 5, which differs from the original bug report. IMO this seems reasonable to match https://github.com/rust-lang/rust/issues/42760.

There should also be test cases for `doc(include = ...)`.

GuillaumeGomez commented:
So I changed a bit how the unindent is handled. For example:

```rust
#[doc = "hello
         second"]
```

Before, we didn't take into account the first line that in this case, we could align the doc based on the next lines indent. It was very misleading and not that great as I discovered while updating this PR. I always assumed it was supposed to be:

```rust
#[doc = "hello
second"]
```

I definitely think it's better this way.

Apart from that change, I also added comments and extended tests to add doc includes.

jyn514 commented:
```
#[doc = "hello
second"]
```
This does not look right. If that's really what the user meant, they can use
```
"hello \
    second"
```
which the lexer will handle without rustdoc having to get involved. I think rustdoc is already trying to be too magic here and we shouldn't special-case things any more than they already are.

jyn514 commented on src/librustdoc/passes/unindent_comments.rs:
```diff
@@ -35,52 +34,69 @@ impl clean::Attributes {
+    // #[doc = "another"]
+    //
+    // In this case, you want "hello! another" and not "hello!  another".
+    let add = if !docs.windows(2).all(|arr| arr[0].kind == arr[1].kind)
```
nit: I think this is easier to read if you expand the boolean logic. Plus it allows short-circuiting.
```suggestion
    let add = if docs.windows(2).any(|arr| arr[0].kind != arr[1].kind)
```

jyn514 commented on src/librustdoc/passes/unindent_comments.rs:
```diff
@@ -35,52 +34,69 @@ impl clean::Attributes {
+    // For example:
+    //
+    // /// hello!
+    // #[doc = "another"]
```
This is a really great comment :) do you have the same example in a test?

GuillaumeGomez replied:
Yep! It's actually the first one I added (even though it doesn't look exactly the same). For example: https://github.com/rust-lang/rust/pull/78400/files#diff-b8ef7a3193cdebf8d2bf582e37fa1689711ed17ef19f6324e25d640f5a796dc0R11

jyn514 commented on src/librustdoc/passes/unindent_comments.rs:
```diff
@@ -35,52 +34,69 @@ impl clean::Attributes {
+    };
+
+    for fragment in docs {
+        let lines: Vec<_> = fragment.doc.lines().collect();
```
nit: this seems a little inefficient, I think you only need the collect for the `is_empty()` call? Instead you could check `fragment.doc.trim().is_empty()` and use `fragment.doc.lines()` for the loop.

GuillaumeGomez replied:
We want to keep the empty lines, they're important in the markdown. :)

jyn514 replied:
Right, I'm saying you can keep the empty lines without having to collect into a vec, like this:
```suggestion
if fragment.doc.lines().count() != 0 {
    fragment.doc = fragment.doc.lines().map(...);
```

jyn514 commented on src/test/rustdoc/unindent.rs:
```diff
@@ -21,3 +23,42 @@ pub struct Example;
+pub struct J;
+
+// @has foo/struct.K.html
+// @matches - '//pre[@class="rust rust-example-rendered"]' '(?m)4 whitespaces!\Z'
```
What does `?m` do? Where are the other lines from the doc-comment?

GuillaumeGomez replied:
`(?m)` checks that it's the beginning of the line. Also, in here I test the content of a codeblock, which is why you don't see the rest. It's easier to check the whitespace handling. :)

jyn514 reviewed (commented):
I like the current behavior and code :) just a few comments on the tests.

jyn514 commented:
> This does not look right.

It turns out I was confused - Guillaume was describing the _old_ behavior, not what rustdoc does after this PR.

pickfire commented on src/test/rustdoc/unindent.rs:
```diff
@@ -0,0 +1,64 @@
+// @matches - '//pre[@class="rust rust-example-rendered"]' '(?m)4 whitespaces!\Z'
+///a
+///
+///    4 whitespaces!
```
Just wondering for the other html sanitization patch, can we do something like a diagram?
```
///    a --> b
```
I feel lazy to go back and check which patch is it.

GuillaumeGomez replied:
You can have HTML in your markdown already, this PR doesn't change anything about the rendering part. :)

jyn514 commented:
r=me with CI passing. Thanks for adding all the tests and comments, this looks a lot better :)

jyn514 commented:
@bors r+

bors commented:
:pushpin: Commit 87f28978c532547df7e8d746ad98f863a202bab0 has been approved by `jyn514`

<!-- @bors r=jyn514 87f28978c532547df7e8d746ad98f863a202bab0 -->
<!-- homu: {"type":"Approved","sha":"87f28978c532547df7e8d746ad98f863a202bab0","approver":"jyn514"} -->

bors commented:
:hourglass: Testing commit 87f28978c532547df7e8d746ad98f863a202bab0 with merge 1d5057d91ad6ad54b899a9b93b5b6b8f1fe73ffa...
<!-- homu: {"type":"BuildStarted","head_sha":"87f28978c532547df7e8d746ad98f863a202bab0","merge_sha":"1d5057d91ad6ad54b899a9b93b5b6b8f1fe73ffa"} -->

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/runs/1339516366)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/runs/1339516366","builder_name":"checks-actions"} -->

GuillaumeGomez commented:
Fetch error, let's retry!

@bors: retry

bors commented:
:hourglass: Testing commit 87f28978c532547df7e8d746ad98f863a202bab0 with merge 83a3d46759cbc915445e3f07addb85a82085b2c7...
<!-- homu: {"type":"BuildStarted","head_sha":"87f28978c532547df7e8d746ad98f863a202bab0","merge_sha":"83a3d46759cbc915445e3f07addb85a82085b2c7"} -->

jyn514 commented:
If this fails because of another zlib error, let's avoid retrying, `@ehuss` is working on a fix for the underlying issue in https://rust-lang.zulipchat.com/#narrow/stream/246057-t-cargo/topic/pinging.20for.20zlib.20error.

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/runs/1339622306)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/runs/1339622306","builder_name":"checks-actions"} -->

GuillaumeGomez commented:
Seeing that other PRs seem to pass, let's give it another try.

@bors: retry
